File: java/Decoupling-Temporal-Services-with-Nexus/solution/src/main/java/compliance/temporal/ComplianceWorkerApp.java
package compliance.temporal;

import compliance.ComplianceChecker;
import compliance.temporal.activity.ComplianceActivityImpl;
import compliance.temporal.workflow.ComplianceWorkflowImpl;
import io.temporal.client.WorkflowClient;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.Worker;
import io.temporal.worker.WorkerFactory;

/**
 * Compliance team's worker — handles Nexus requests from Payments.
 * Task queue: "compliance-risk"
 *
 * Registers three things:
 *   1. ComplianceWorkflowImpl — the workflow that wraps the activity
 *   2. ComplianceActivityImpl — the activity that runs the checker
 *   3. ComplianceNexusServiceImpl — the Nexus handler that launches the workflow
 */
public class ComplianceWorkerApp {

    public static void main(String[] args) {
        // C — Connect to Temporal
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClient client = WorkflowClient.newInstance(service);

        // R — Create factory and worker
        WorkerFactory factory = WorkerFactory.newInstance(client);
        Worker worker = factory.newWorker("compliance-risk");

        // W — Wire workflow, activity, and Nexus handler
        worker.registerWorkflowImplementationTypes(ComplianceWorkflowImpl.class);
        worker.registerActivitiesImplementations(new ComplianceActivityImpl(new ComplianceChecker()));
        worker.registerNexusServiceImplementation(new ComplianceNexusServiceImpl());

        // L — Launch
        factory.start();

        System.out.println("=========================================================");
        System.out.println("  Compliance Worker started on: compliance-risk");
        System.out.println("  Registered: ComplianceWorkflow, ComplianceActivity,");
        System.out.println("              ComplianceNexusServiceImpl (sync handler)");
        System.out.println("  Waiting for Nexus requests from Payments team...");
        System.out.println("=========================================================");
    }
}

File: kotlin/decouple-monolith/exercise/src/main/kotlin/compliance/temporal/ComplianceWorkerApp.kt
package compliance.temporal

import compliance.ComplianceChecker
import compliance.temporal.activity.ComplianceActivityImpl
import compliance.temporal.workflow.ComplianceWorkflowImpl
import io.temporal.client.WorkflowClient
import io.temporal.client.WorkflowClientOptions
import io.temporal.serviceclient.WorkflowServiceStubs
import io.temporal.worker.WorkerFactory

/**
 * The Compliance team's Worker. This process is the only place their code runs.
 *
 * The connection and the Worker itself are set up for you. What is missing is the part
 * that decides what this Worker can actually do: see TODO 7 below.
 */
fun main() {
    // Client (newInstance): connect, scoped to the Compliance team's own Namespace.
    val service = WorkflowServiceStubs.newLocalServiceStubs()
    val client = WorkflowClient.newInstance(
        service,
        WorkflowClientOptions.newBuilder()
            .setNamespace("compliance-namespace")
            .build(),
    )

    // Worker (newWorker): poll the Task Queue the Endpoint will point at. This value
    // MUST match the --target-task-queue you give the Nexus Endpoint, or the Endpoint
    // routes work to a queue nobody is polling.
    val factory = WorkerFactory.newInstance(client)
    val taskQueue = "compliance-risk"
    val worker = factory.newWorker(taskQueue)

    // ── TODO 7 ───────────────────────────────────────────────────────────────────
    // Register (register*): a Worker only handles work it has been told about. This one
    // is responsible for three things, each a separate call on `worker`:
    //
    //   1. The Workflow that carries out a compliance check.
    //   2. The Activity that Workflow calls to run the actual risk assessment. This one
    //      is not registered by type: the Worker needs a constructed instance, and that
    //      instance needs a ComplianceChecker to do its work.
    //   3. The Nexus handler, which is what makes this team callable by Payments at all.
    //      Leave this one out and the Worker still starts, Payments' calls just go
    //      unanswered forever. That is the failure you are guarding against.
    //
    // Each call name starts with `register`.
    //
    // Stuck? "Ask AI" on https://docs.temporal.io:
    //   "what do I register on a Java Worker that handles Nexus Operations?"


    // Start (start): begin polling.
    factory.start()

    println("=========================================================")
    println("  Compliance Worker started on: $taskQueue")
    println("  Namespace: compliance-namespace")
    println("=========================================================")
}
